library/src/main/java/com/vinted/coper/CoperFragment.kt

```diff
@@ -44,6 +49,12 @@ internal class CoperFragment : Fragment() {
         }
     }
 
+    @VisibleForTesting
+    internal suspend fun waitUntilRequestStart() {
+        triggerPermissionRequest = CompletableDeferred()
+        triggerPermissionRequest?.await()
+    }
+
```

Comment:
This code does not look safe in race-condition regards. One case is, each time you are calling `waitUntilRequestStart()`, you basically lock other suspend functions forever. It may be OK for tests. Other code increases chances of race-condition as well:
```
            if (value != null) triggerPermissionRequest?.complete(Unit)
            triggerPermissionRequest = null
```

My recommendation would be considered to expose `permissionRequestState` event stream to the tests which could look like:
```
fixture.permissionRequestStateFlow.filter { /* check for correct event */ ).first()
```
You also will have the flexibility to do more things.